### DISCUSSION TITLE
Uncaught TypeError: can't access property "_x_refs", root is undefined

### DISCUSSION THREAD

**Discussion by StoyanDimitrov** [alpinejs/alpine#3199] [7. Bugs (archived)] [answered]
Probably related to #1568, #2228, and #3003. The minimal failing code is
```html
<template x-data="{}" x-if="true">
  <div x-ref="foo">Contents</div>
</template>
```

This one works as expected
```html
<div x-data="{}">
  <template x-if="true">
    <div x-ref="foo">Inside</div>
  </template>
</div>
```

**Comment by ekwoka** [comment, marked as answer]
This is not a bug. This is expected behavior.

The `x-ref` will not be evaluated inside of the `x-data` since it won't be inside the `x-data`

**Comment by StoyanDimitrov** [comment]
Can you elaborate a bit?

**Reply by ekwoka** [reply]
The properties would be evaluated in the context of that elements position.

the x-if copies the templates content to be directly after the template.

In your example, it would then no longer be inside the x-data.

**Comment by StoyanDimitrov** [comment]
Is that changed between v2 and v3?

**Reply by ekwoka** [reply]
Idk that was a while ago.

It's not an expected behavior in v3 though for sure.

You should wrap the x-if completely in the x-data
